comms/src/connection_manager/protocol.rs

```diff
@@ -0,0 +1,171 @@
+
+        let message_header = MessageHeader {
+            message_type: ControlServiceMessageType::EstablishConnection,
+        };
+        let msg = Message::from_message_format(message_header, msg).map_err(ConnectionManagerError::MessageError)?;
+        let body = msg.to_binary().map_err(ConnectionManagerError::MessageFormatError)?;
+
+        // Encrypt body
+        let encrypted_body = self.encrypt_body(&node_identity, body)?;
+
+        let envelope = MessageEnvelope::construct(
+            node_identity,
+            NodeDestination::NodeId(self.peer.node_id.clone()),
+            encrypted_body,
+            MessageFlags::ENCRYPTED,
+        )
+        .map_err(ConnectionManagerError::MessageError)?;
+
+        control_conn
+            .send(envelope.into_frame_set())
+            .map_err(ConnectionManagerError::ConnectionError)?;
+
+        Ok(())
+    }
+
+    fn encrypt_body(&self, identity: &Arc<CommsNodeIdentity>, body: Frame) -> Result<Frame> {
+        let ecdh_shared_secret = CommsPublicKey::shared_secret(&identity.secret_key, &self.peer.public_key).to_vec();
+        let ecdh_shared_secret_bytes: [u8; 32] = ByteArray::from_bytes(&ecdh_shared_secret)
+            .map_err(ConnectionManagerError::SharedSecretSerializationError)?;
+        Ok(chacha20::encode(&body, &ecdh_shared_secret_bytes))
+    }
+
+    fn open_inbound_peer_connection(
+        &self,
+        connections: &Arc<LivePeerConnections>,
+    ) -> Result<(Arc<PeerConnection>, EstablishConnection)>
+    {
+        let (secret_key, public_key) =
+            CurveEncryption::generate_keypair().map_err(ConnectionManagerError::CurveEncryptionGenerateError)?;
+
+        let address = connections.establish_connection(ConnectionDirection::Inbound {
+            node_id: self.peer.node_id.clone(),
+            secret_key,
+        })?;
+
+        let node_identity = CommsNodeIdentity::global().ok_or(ConnectionManagerError::GlobalNodeIdentityNotSet)?;
+
+        let connection = connections
+            .get_connection(&self.peer.node_id)
+            .ok_or(ConnectionManagerError::PeerConnectionNotFound)?;
+
+        Ok((connection, EstablishConnection {
+            address,
+            control_service_address: node_identity.control_service_address.clone(),
+            public_key: node_identity.identity.public_key.clone(),
+            node_id: node_identity.identity.node_id.clone(),
+            server_key: public_key,
+        }))
+    }
+
+    fn establish_control_service_connection(
+        &self,
+        connections: &Arc<LivePeerConnections>,
+        server_public_key: CurvePublicKey,
+    ) -> Result<EstablishedConnection>
+    {
+        let context = connections.borrow_context();
+        let config = &connections.config;
+
+        // TODO: Set net address stats and try all net addresses before giving up
+        let address = self.peer.addresses.addresses[0].clone().as_net_address();
```

Comment:
I think you can keep calling `get_best_net_address` until it returns an error result, then all connection attempts would have been used

Reply:
Problem is the mutable reference

Reply:
PeerManager does expose the `get_best_net_address` and should be usable